### PR TITLE
Use refresh tokens

### PR DESCRIPTION
**Pull request by jeffyip**
Updated this demo app to use refresh tokens.

### CODE REVIEW

File: app.rb [hunk 1]
```diff
@@ -3,27 +3,51 @@
...
-  File.open("refresh_token.txt", "w") do |file|
-    file.write(@transaction_tokens.refresh_token)
+  # Check if refresh token exists for given UID
+  if File.file?("#{@uid}_refresh_token.txt")
+    # read the refresh token from disk
+    @refresh_token = File.open("#{@uid}_refresh_token.txt").read
+    
+    begin
+      @transaction_tokens = Berbix::Tokens.from_refresh(@refresh_token)
+      @data = client.fetch_transaction(@transaction_tokens) # needed to force library to refresh the refresh token
```
Comment (jeffyip) [inline app.rb]: @ericlevine I'm mainly concerned to see if the above 2 lines are correct and recommended.  It seems to me that the call to `client.fetch_transaction(@transaction_tokens)` is needed to `refresh` the `refresh_token`, is that correct?  But the data returned from `fetch_transaction` isn't really needed at least in this use-case.

Comment (ericlevine) [reply app.rb]: You can call `refresh_tokens` to get a new set of fresh tokens if your goal is to get a new `client_token` - I'm guessing that's what you meant, right? You don't need to fetch a new `refresh_token` given that it's long-lived.

File: app.rb [hunk 2]
```diff
@@ -3,27 +3,51 @@
...
-    template_key: berbix_config['template_key'], # Template key for this transaction
-  )
-
-  # To avoid dealing with a DB for this demo, we'll write the refresh token to a file
-  File.open("refresh_token.txt", "w") do |file|
-    file.write(@transaction_tokens.refresh_token)
+  # Check if refresh token exists for given UID
+  if File.file?("#{@uid}_refresh_token.txt")
+    # read the refresh token from disk
+    @refresh_token = File.open("#{@uid}_refresh_token.txt").read
```
Comment (ericlevine) [inline app.rb]: `/path?uid=../../../etc/passwd`. There's got to be a way to get `File.open` to ignore the second part of a string...

File: app.rb (hunk 1 above)
Comment (jeffyip) [reply app.rb:36]: Ah I tried that here and I got this error:
Screen Shot 2020-11-16 at 9 28 03 PM

I'll try to debug tomorrow.

File: app.rb [hunk 3]
```diff
@@ -3,42 +3,65 @@
...
+    # read the refresh token from disk
+    @refresh_token = File.open("#{@uid.to_i}_refresh_token.txt").read
+    @transaction_tokens = Berbix::Tokens.from_refresh(@refresh_token)
+    @transaction_data = client.fetch_transaction(@transaction_tokens)
+    if @transaction_data && !@transaction_data["action"].nil?
+      logger.info @transaction_data["action"]
+      @status = "completed"
+    else
+      @status = "incomplete"
+    end
```
Comment (jeffyip) [inline app.rb]: @ericlevine what do you think about taking this approach -- essentially calling `fetch_transaction` to introspect on the status of the transaction?

Comment (ericlevine) [reply app.rb]: Is your goal with this to figure out whether the transaction has been completed or not? I think knowing what the ultimate purpose is will help with evaluating whether this is the right approach.

Comment (jeffyip) [reply app.rb]: Yes, the goal is to illustrate how one would introspect on the status of the transaction if they wanted to render something to the user in accordance with their current transaction state.

Comment (jeffyip) [reply app.rb]: Ultimately want to illustrate the "best practices" here.

Comment (jeffyip) [reply app.rb]: Chatted with @jnnnthnn and landed on this:
- don't encourage this approach, and instead recommend that customers use webhooks and/or oncomplete callbacks to update the state of the transaction, and make sure that customers update the `completed_at` time for a `refresh_token` in their own database
- encourage customers to rely on the client to determine the state of the transaction on load
- this approach could be useful for a backfill if the webhooks/callback approach failed for some reason

**Pull request merged**